### DISCUSSION TITLE
How to convert .bmp file to byte array?

### DISCUSSION THREAD

**Discussion by trykof** [CosmosOS/Cosmos#1897] [Q&A]
My OS doesn't load .bmp file. Then, how to convert .bmp file to byte array?

**Comment by alf1e** [comment]
I found this for converting images to byte arrays https://stackoverflow.com/questions/3801275/how-to-convert-image-to-byte-array

**Reply by trykof** [reply]
Image drawing, but incorrectly. Why?

Image in OS:
![изображение_2021-08-05_202510](https://user-images.githubusercontent.com/66679206/128393968-acebbba6-62f0-469e-b152-d32a2b759ca4.png)

Original Image:
![lightning-os-logo](https://user-images.githubusercontent.com/66679206/128394034-8b31b2d2-b778-4c41-b5b0-2af1d66ab2ae.png)

**Reply by trykof** [reply]
OK, I'm partially fixed the problem. There is only one thing left. Why are the colors not displayed correctly?

**Reply by alf1e** [reply]
I'm not 100% sure, might have to google that

**Reply by alf1e** [reply]
> OK, I'm partially fixed the problem. There is only one thing left. Why are the colors not displayed correctly?

Maybe fiddle with the format

**Reply by trykof** [reply]
Why colors so weird... Lots of yellow...

![изображение_2021-08-05_205103](https://user-images.githubusercontent.com/66679206/128397217-d5901826-47b0-4a4f-9dd1-cf22506a9d6e.png)

**Reply by alf1e** [reply]
Ahh, the image is RGB and byte array iS BGR so need to convert them

**Reply by alf1e** [reply]
https://stackoverflow.com/questions/19187737/converting-a-bgr-bitmap-to-rgb try editing this

**Reply by trykof** [reply]
Code in this article, doesn't work. It only got worse.

![изображение_2021-08-05_215551](https://user-images.githubusercontent.com/66679206/128405418-1b5127eb-775e-43cd-bb5f-a7ea56fd9a15.png)

Maybe error in a code?  I'm modified it. 

`public static void RGBtoBGR(Bitmap bmp)
        {
            BitmapData data = bmp.LockBits(new Rectangle(0, 0, bmp.Width, bmp.Height),
                                           ImageLockMode.ReadWrite, bmp.PixelFormat);

            IntPtr ptr = data.Scan0;

            int bytes = Math.Abs(data.Stride) * bmp.Height;

            byte[] rgbValues = new byte[bytes];

            Marshal.Copy(ptr, rgbValues, 0, bytes);

            for (int i = 0; i < bytes; i += 3)
            {
                byte dummy = rgbValues[i];
                rgbValues[i] = rgbValues[i + 2];
                rgbValues[i + 2] = dummy;
            }

           Marshal.Copy(rgbValues, 0, ptr, bytes);

           bmp.UnlockBits(data);

            bmp.Save("input.bmp");
        }`

**Reply by alf1e** [reply]
I'm not sure then

**Reply by trykof** [reply]
This bmp file after convertation to BGR pixel format.

![input](https://user-images.githubusercontent.com/66679206/128405928-65875e22-3204-471b-ad2b-c7ace0f902fd.png)

**Reply by alf1e** [reply]
You were meant to render straight from the byte array

**Reply by trykof** [reply]
Yes, I'm doing this

**Reply by trykof** [reply]
To render image I'm using this code:

`byte[] img = new byte[] { ..bmp code here... };
Bitmap bmp = new Bitmap(img, ColorOrder.BGR);
Canvas g = FullScreenCanvas.GetFullScreenCanvas();
g.DrawImage(bmp, 10, 10);`

**Reply by alf1e** [reply]
Try changing ColorOrder to RGB

**Reply by alf1e** [reply]
> Try changing ColorOrder to RGB

ColorOrder.RGB

**Reply by trykof** [reply]
Nothing. Image doesn't draw. Black screen.

**Reply by alf1e** [reply]
> Nothing. Image doesn't draw. Black screen.

RGBA or BGRA

**Reply by trykof** [reply]
I need use Canvas.DrawImageAlpha()? In enum ColorOrder no RGBA or BGRA.
